update docs related to partition table dynamic mode (#5810)
pingcap/docs: partitioned-table.md

The `tidb_enable_list_partition` environment variable controls whether to enable the partitioned table feature. If this variable is set to `OFF`, the partition information will be ignored when a table is created, and this table will be created as a normal table.

This variable is only used in table creation. After the table is created, modify this variable value takes no effect. For details, see [system variables](/system-variables.md#tidb_enable_list_partition-new-in-v50).

### Dynamic mode

> **Warning:**
>
> This is still an experimental feature. It is **NOT** recommended that you use it in the production environment.

> There are two modes for TiDB accessing partitioned tables: `dynamic` and `static`. Currently, the `static` mode is used by default. If you want to enable `dynamic` mode, you need to manually set `tidb_partition_prune_mode` to `dynamic`.

Reviewer: ```suggestion
TiDB accesses partitioned tables in one of the two modes: `dynamic` mode and `static` mode. Currently, `static` mode is used by default. If you want to enable `dynamic` mode, you need to manually set the `tidb_partition_prune_mode` variable to `dynamic`.
```

{{< copyable "sql" >}}

```sql
set @@session.tidb_partition_prune_mode = 'dynamic'
```

> In `static` mode, TiDB accesses each partition separately with multiple operators, and then merges the results through `Union`. The following example is a simple read operation, where you can find that TiDB merges the results of two corresponding partitions with `Union`:

Reviewer: ```suggestion
In `static` mode, TiDB accesses each partition separately using multiple operators, and then merges the results using `Union`. The following example is a simple read operation where TiDB merges the results of two corresponding partitions using `Union`:
```

+------------------------------+----------+-----------+------------------------+--------------------------------+
7 rows in set (0.00 sec)
```

In `dynamic` mode, each operator supports direct access to multiple partitions, so TiDB no longer uses `Union`.

{{< copyable "sql" >}}

```sql
mysql> set @@session.tidb_partition_prune_mode = 'dynamic';
Query OK, 0 rows affected (0.00 sec)

mysql> explain select * from t1 where id < 150;
+-------------------------+----------+-----------+-----------------+--------------------------------+
| id                      | estRows  | task      | access object   | operator info                  |
+-------------------------+----------+-----------+-----------------+--------------------------------+
| TableReader_7           | 3323.33  | root      | partition:p0,p1 | data:Selection_6               |
| └─Selection_6           | 3323.33  | cop[tikv] |                 | lt(test.t1.id, 150)            |
|   └─TableFullScan_5     | 10000.00 | cop[tikv] | table:t1        | keep order:false, stats:pseudo |
+-------------------------+----------+-----------+-----------------+--------------------------------+
3 rows in set (0.00 sec)
```

> From the above query results, it can be seen that the `Union` opearator in the execution plan disappears while the partition pruning is still in effect and the execution plan only accesses `p0` and `p1`.

Reviewer: ```suggestion
From the above query results, you can see that the `Union` operator in the execution plan disappears while the partition pruning still takes effect and the execution plan only accesses `p0` and `p1`.
```

> `dynamic` mode makes execution plans more simple and clear. Omitting the Union operation can improve the execution efficiency and avoid the problem of Union concurrent execution. In addition, `dynamic` mode also solves two problems that cannot be solved in `static` mode:

Reviewer: ```suggestion
`dynamic` mode makes execution plans simpler and clearer. Omitting the Union operation can improve the execution efficiency and avoid the problem of Union concurrent execution. In addition, `dynamic` mode also solves two problems that cannot be solved in `static` mode:
```

> + Do not use Plan Cache (see example 1 and 2)
> + Do not use IndexJoin type (see example 3 and 4)

Reviewer: ```suggestion
+ Plan Cache cannot be used. (See example 1 and 2)
+ Execution plans with IndexJoin cannot be used. (See example 3 and 4)
```

> **Example 1**：The following example enables Plan Cache function in the configuration file and executes the same query twice in `static` mode:

Reviewer: ```suggestion
**Example 1**：In the following example, the Plan Cache feature is enabled in the configuration file and the same query is executed twice in `static` mode:
```

{{< copyable "sql" >}}

```sql
mysql> set @a=150;
Query OK, 0 rows affected (0.00 sec)

mysql> set @@tidb_partition_prune_mode = 'static';
Query OK, 0 rows affected (0.00 sec)

mysql> prepare stmt from 'select * from t1 where id < ?';
Query OK, 0 rows affected (0.00 sec)

mysql> execute stmt using @a;
Empty set (0.00 sec)

mysql> execute stmt using @a;
Empty set (0.00 sec)

> -- In static mode, when the same query is executed twice, the cache cannot be hit the second time

Reviewer: ```suggestion
-- In static mode, when the same query is executed twice, the cache cannot be hit at the second time.
```

mysql> select @@last_plan_from_cache;
+------------------------+
| @@last_plan_from_cache |
+------------------------+
|                      0 |
+------------------------+
1 row in set (0.00 sec)
```

> `last_plan_from_cache` variable can display whether the last query hit the Plan Cache. As can be seen from above, in `static` mode, even if the same query is executed multiple times on the partitioned table, the Plan Cache will not be hit.

Reviewer: ```suggestion
The `last_plan_from_cache` variable can show whether the last query hits the Plan Cache or not. From example 1, you can see that in `static` mode, even if the same query is executed multiple times on the partitioned table, the Plan Cache will not be hit.
```

> The `last_plan_from_cache` variable can show whether the last query hits the Plan Cache or not. From example 1, you can see that in `static` mode, even if the same query is executed multiple times on the partitioned table, the Plan Cache is not be hit.

Reviewer: ```suggestion
The `last_plan_from_cache` variable can show whether the last query hits the Plan Cache or not. From example 1, you can see that in `static` mode, even if the same query is executed multiple times on the partitioned table, the Plan Cache is not hit.
```

> **Example 2**: The following example performs the same operations as example 1 in `dynamic` mode:

Reviewer: ```suggestion
**Example 2**: In the following example, the same operations are performed in `dynamic` mode as are done in example 1:
```

{{< copyable "sql" >}}

```sql
mysql> set @@tidb_partition_prune_mode = 'dynamic';
Query OK, 0 rows affected (0.00 sec)

mysql> prepare stmt from 'select * from t1 where id < ?';
Query OK, 0 rows affected (0.00 sec)

mysql> execute stmt using @a;
Empty set (0.00 sec)

mysql> execute stmt using @a;
Empty set (0.00 sec)

> -- in dynamic mode, the cache can be hit the second time

Reviewer: ```suggestion
-- In dynamic mode, the cache can be hit at the second time.
```

mysql> select @@last_plan_from_cache;
+------------------------+
| @@last_plan_from_cache |
+------------------------+
|                      1 |
+------------------------+
1 row in set (0.00 sec)
```

> As can be seen from above, in `dynamic` mode, querying partitioned tables can hit the Plan Cache.

Reviewer: ```suggestion
From example 2, you can see that in `dynamic` mode, querying the partitioned table hits the Plan Cache.
```

> **Example 3**: The following example tries the execution plan with IndexJoin query in `static` mode:

Reviewer: ```suggestion
**Example 3**: In the following example, a query is performed in `static` mode using the execution plan with IndexJoin:
```

|   │   └─TableFullScan_22       | 10000.00 | cop[tikv] | table:t1, partition:p1 | keep order:false, stats:pseudo                 |
|   ├─TableReader_30             | 9990.00  | root      |                        | data:Selection_29                              |
|   │ └─Selection_29             | 9990.00  | cop[tikv] |                        | not(isnull(test.t1.id))                        |
|   │   └─TableFullScan_28       | 10000.00 | cop[tikv] | table:t1, partition:p2 | keep order:false, stats:pseudo                 |
|   └─TableReader_36             | 9990.00  | root      |                        | data:Selection_35                              |
|     └─Selection_35             | 9990.00  | cop[tikv] |                        | not(isnull(test.t1.id))                        |
|       └─TableFullScan_34       | 10000.00 | cop[tikv] | table:t1, partition:p3 | keep order:false, stats:pseudo                 |
+--------------------------------+----------+-----------+------------------------+------------------------------------------------+
17 rows in set, 1 warning (0.00 sec)
```

> As can be seen from above, even if the hint of `TIDB_INLJ` is used, the query with partitioned table can not select the execution plan with IndexJoin.

Reviewer: ```suggestion
From example 3, you can see that even if the `TIDB_INLJ` hint is used, the query on the partitioned table cannot select the execution plan with IndexJoin.
```

> **Example 4**: The following example tries the execution plan with IndexJoin query in `dynamic` mode:

Reviewer: ```suggestion
**Example 4**: In the following example, the query is performed in `dynamic` mode using the execution plan with IndexJoin:
```

| └─IndexLookUp_10(Probe)         | 1.25     | root      | partition:all          |                                                                                                                     |
|   ├─Selection_9(Build)          | 1.25     | cop[tikv] |                        | not(isnull(test.t1.id))                                                                                             |
|   │ └─IndexRangeScan_7          | 1.25     | cop[tikv] | table:t1, index:id(id) | range: decided by [eq(test.t1.id, test.t2.id)], keep order:false, stats:pseudo                                      |
|   └─TableRowIDScan_8(Probe)     | 1.25     | cop[tikv] | table:t1               | keep order:false, stats:pseudo                                                                                      |
+---------------------------------+----------+-----------+------------------------+---------------------------------------------------------------------------------------------------------------------+
8 rows in set (0.00 sec)
```

> As can be seen from above, in `dynamic` mode, the plan with IndexJoin is selected when you execute query.

Reviewer: ```suggestion
From example 4, you can see that in `dynamic` mode, the execution plan with IndexJoin is selected when you execute the query.
```
